src/ncp/posix/netif.cpp

```diff
@@ -289,9 +354,137 @@ void Netif::Clear(void)
+#endif
+
     mNetifIndex = 0;
     mIp6UnicastAddresses.clear();
     mIp6MulticastAddresses.clear();
 }
 
+#if OTBR_NETIF_USE_MLD_MONITOR
+static const otIp6Address kMLDv2MulticastAddress = {
+    {{0xff, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x16}}};
+static const otIp6Address kAllRouterLocalMulticastAddress = {
+    {{0xff, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02}}};
+
+static bool IsMulAddrFiltered(const otIp6Address &aAddr)
+{
+    return memcmp(&aAddr, &kMLDv2MulticastAddress, sizeof(otIp6Address)) == 0 ||
+           memcmp(&aAddr, &kAllRouterLocalMulticastAddress, sizeof(otIp6Address)) == 0;
+}
+
+otbrError Netif::InitMldListener(void)
+{
+    otbrError        error = OTBR_ERROR_NONE;
+    struct ipv6_mreq mreq6;
+
+    mMldFd = SocketWithCloseExec(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6, kSocketNonBlock);
+    VerifyOrExit(mMldFd != -1, error = OTBR_ERROR_ERRNO);
+
+    mreq6.ipv6mr_interface = mNetifIndex;
+    memcpy(&mreq6.ipv6mr_multiaddr, kMLDv2MulticastAddress.mFields.m8, sizeof(kMLDv2MulticastAddress.mFields.m8));
+
+    VerifyOrExit(setsockopt(mMldFd, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq6, sizeof(mreq6)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+#ifdef __linux__
+    VerifyOrExit(setsockopt(mMldFd, SOL_SOCKET, SO_BINDTODEVICE, mNetifName.c_str(),
+                            static_cast<socklen_t>(mNetifName.length())) == 0,
+                 error = OTBR_ERROR_ERRNO);
+#endif
+
+exit:
+    return error;
+}
+
+void Netif::ProcessMldEvent(void)
+{
+    const size_t        kMaxMldEvent = 8192;
+    uint8_t             buffer[kMaxMldEvent];
+    ssize_t             bufferLen = -1;
+    struct sockaddr_in6 srcAddr;
+    socklen_t           addrLen  = sizeof(srcAddr);
+    bool                fromSelf = false;
+    Mldv2Header        *hdr      = reinterpret_cast<Mldv2Header *>(buffer);
+    size_t              offset;
+    uint8_t             type;
+    struct ifaddrs     *ifAddrs = nullptr;
+    char                addressString[INET6_ADDRSTRLEN + 1];
+
+    bufferLen = recvfrom(mMldFd, buffer, sizeof(buffer), 0, reinterpret_cast<sockaddr *>(&srcAddr), &addrLen);
+    VerifyOrExit(bufferLen > 0);
+
+    type = buffer[0];
+    VerifyOrExit(type == kIcmpv6Mldv2Type && bufferLen >= static_cast<ssize_t>(sizeof(Mldv2Header)));
+
+    // Check whether it is sent by self
+    VerifyOrExit(getifaddrs(&ifAddrs) == 0);
+    for (struct ifaddrs *ifAddr = ifAddrs; ifAddr != nullptr; ifAddr = ifAddr->ifa_next)
+    {
+        if (ifAddr->ifa_addr != nullptr && ifAddr->ifa_addr->sa_family == AF_INET6 &&
+            strncmp(mNetifName.c_str(), ifAddr->ifa_name, IFNAMSIZ) == 0)
+        {
+            struct sockaddr_in6 *addr6 = reinterpret_cast<struct sockaddr_in6 *>(ifAddr->ifa_addr);
+
+            if (memcmp(&addr6->sin6_addr, &srcAddr.sin6_addr, sizeof(in6_addr)) == 0)
+            {
+                fromSelf = true;
+                break;
```

Comment:
nit: maybe fromSelf is not needed if using `ExitNow();` here.

Reply:
It seems we cannot remove it because the condition is to verify that at least one address matches and only exit when all addresses don't match.